openssl-book.tex

```diff
@@ -27,14 +28,259 @@
+      \item{Ciphersuite selection (client vs server preference)}
+      \item{Key Exchange Mechanisms}
+      \begin{outline}
+        \item{RSA}
+        \item{DHE}
+        \item{ECDHE (covering some basics of curve types: P-256, X25519 etc)}
+        \item{SRP}
+        \item{PSK}
+      \end{outline}
+      \item{Authentication}
+      \begin{outline}
+        \item{RSA}
+        \item{ECDSA}
+        \item{EdDSA? (future)}
+      \end{outline}
+      \item{Encryption}
+      \begin{outline}
+        \item{AES}
+        \item{Camellia}
+        \item{ChaCha}
+        \item{etc}
+      \end{outline}
+      \item{MAC/AEAD}
+    \end{outline}
+    \item{Chapter: Basic Operation}
+    \begin{outline}
+      \item{The read and write BIOs}
+      \item{Alerts}
+      \item{Version Negotiation}
+      \item{SSL\_read, SSL\_write and SSL\_get\_error}
+      \begin{outline}
+        \item{Non-blocking IO}
+        \item{Pending data}
+      \end{outline}
+      \item{Shutting down}
+      \item{Client Authentication}
+      \item{Renegotiation}
+      \item{Compression}
+      \item{SSL BIO}
+      \item{Exporting secrets}
+    \end{outline}
+    \item{Chapter: Sessions}
+    \begin{outline}
+      \item{Resumption handshakes}
+      \item{Simple sessions and session files}
+      \item{Session tickets}
+      \item{Session caches}
+    \end{outline}
+    \item{Chapter: Configuration}
+    \begin{outline}
+      \item{Setting options and modes}
+      \begin{outline}
+        \item{Some common options/modes}
+        \begin{outline}
+          \item{SSL\_MODE\_AUTO\_RETRY}
+          \item{SSL\_MODE\_RELEASE\_BUFFERS}
+        \end{outline}
+      \end{outline}
+      \item{Signature Algorithms}
+      \item{Supported Groups}
+      \item{Configuration using SSL\_CONF}
+      \item{Security levels and the security callbacks}
+    \end{outline}
+    \item{Chapter: DTLS}
+    \begin{outline}
+      \item{Key differences with TLS}
+      \item{Transports}
+      \begin{outline}
+        \item{UDP}
+        \item{SCTP}
+        \item{MTU issues}
+      \end{outline}
+      \item{Retransmissions and the DTLS timer}
+      \item{Listening for connections and cookies}
+    \end{outline}
+    \item{Chapter: TLSv1.3}
+    \item{Chapter: Debugging Connection Failures}
+    \item{Chapter: Advanced Extensions (Advanced Topic)}
+    \begin{outline}
+      \item{SNI}
+      \item{ALPN and NPN}
+      \item{SRTP}
+      \item{EC point formats}
+      \item{Extended Master Secret}
+      \item{Encrypt-Then-MAC}
+      \item{OCSP in SSL/TLS}
+      \item{Certificate Transparency in SSL/TLS}
+      \item{Custom extensions}
+    \end{outline}
+    \item{Chapter: DANE (Advanced Topic)}
+    \item{Chapter: Optimisation (Advanced Topic)}
+    \begin{outline}
+      \item{Multiblock}
+      \item{Async}
+      \item{Pipelining}
+      \item{Fragment sizes}
+      \item{Read ahead}
+    \end{outline}
+  \end{outline}
+  \item{Part: Cryptography}
```

Comment:
I'm a little torn if we want to put crypto before SSL/TLS.

Reply:
It's a good question. From a strict layering perspective libssl depends on libcrypto, so in that way it makes sense for crypto to come first. However, IMO more people are using libssl than (just) libcrypto so it made more sense to me for ssl stuff to come first.